### PR TITLE
Fix NPE when calling closures within resource functions

### LINKED ISSUE #30022: Using service argument in query operation results in bad sad error
The problematic code line is,

```ballerina
TLift[] tLifts = from var l in liftTable where l.status == status select l;
```

If I change it to the below it works,

```ballerina
TLift[] tLifts = from var l in liftTable where l.status == "OPEN" select l;
```

The below is the complete program.

```ballerina
import ballerina/graphql;

type TLift record {|
    readonly string id;
    string name;
    string status;
    int capacity;
    boolean night;
    int elevationgain;
    TTrail trailAccess;
|};

type TTrail record {|
    string id;
    string name;
    string status;
    string difficulty;
    boolean groomed;
    boolean trees;
    boolean night;
    // Lift[] accessedByLifts;
|};

table<TLift> key(id) liftTable = table [
    { id: "l1", name: "Lift1", status: "OPEN", capacity: 10, night: false, elevationgain: 20, trailAccess: {id: "t1", name: "tail1", status: "OPEN", difficulty: "HARD", groomed: true, trees: false, night: false}}
];

service /graphql on new graphql:Listener(9000) {
    resource function get allLifts(string status) returns Lift {
        TLift[] tLifts = from var l in liftTable where l.status == status select l;
        return new Lift(tLifts[0]);  
    }
}

service class Lift {

    private TLift tLift;

    function init(TLift tLift) {
        self.tLift = tLift;
    }

    resource function get id () returns string {
        return self.tLift.id;
    }

    resource function get name () returns string {
        return self.tLift.name;
    }

    resource function get status () returns string {
        return self.tLift.status;
    }

    resource function get capacity () returns int {
        return self.tLift.capacity;
    }

    resource function get night () returns boolean {
        return self.tLift.night;
    }

    resource function get elevationgain () returns int {
        return self.tLift.elevationgain;
    }

    resource function get trailAccess () returns Trail {
        return new Trail(self.tLift.trailAccess);
    }
}

service class Trail {

    private TTrail tTrail;

    function init(TTrail tTrail) {
        self.tTrail = tTrail;
    }

    resource function get id () returns string {
        return self.tTrail.id;
    }

    resource function get name () returns string {
        return self.tTrail.name;
    }

    resource function get status () returns string {
        return self.tTrail.status;
    }

    resource function get difficulty () returns string {
        return self.tTrail.difficulty;
    }

    resource function get groomed () returns boolean {
        return self.tTrail.groomed;
    }

    resource function get trees () returns boolean {
        return self.tTrail.trees;
    }

    resource function get night () returns boolean {
        return self.tTrail.night;
    }
}
```

### LINKED ISSUE #30111: NPE when using closures inside a resource functions
**Description:**
Closures aren't working inside resource functions.  ClosureDesugar:collectClosureMapSymbols() is not collecting enclosing symbols. 

```
    java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BirVariableOptimizer.checkForFreeVars(BirVariableOptimizer.java:83)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BirVariableOptimizer.reuseVars(BirVariableOptimizer.java:75)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BirVariableOptimizer.visit(BirVariableOptimizer.java:62)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:380)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BirVariableOptimizer.lambda$visit$2(BirVariableOptimizer.java:53)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BirVariableOptimizer.visit(BirVariableOptimizer.java:53)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRTypeDefinition.accept(BIRNode.java:489)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BirVariableOptimizer.lambda$visit$0(BirVariableOptimizer.java:47)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BirVariableOptimizer.visit(BirVariableOptimizer.java:47)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:83)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BirVariableOptimizer.optimizeNode(BirVariableOptimizer.java:42)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:84)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:269)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:253)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:163)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:397)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:333)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:147)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:133)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:107)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:152)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:106)
        at org.ballerinalang.test.BCompileUtil.jBallerinaBackend(BCompileUtil.java:164)
        at org.ballerinalang.test.BCompileUtil.compile(BCompileUtil.java:75)
        at org.ballerinalang.test.types.stream.SandboxTest.setup(SandboxTest.java:38)
```

**Steps to reproduce:**

```ballerina
import ballerina/lang.test;

listener test:MockListener echoEP = new(9090);

service /graphql on echoEP {

    resource function get allLifts(string status) returns string {
        var addFunc = function (int funcInt) returns (string) {
            return status;
        };
        return addFunc(1);
    }

}
```

**Affected Versions:**
slp-alpha5

**OS, DB, other environment details and versions:**
n/a

**Related Issues (optional):**
- https://github.com/ballerina-platform/ballerina-lang/issues/30022

### PR DESCRIPTION
**Pull request by grainier**
## Purpose
$title

Fixes #30022
Fixes #30111

## Approach
n/a

## Samples
ref:tests

## Remarks
n/a

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+listener test:MockListener echoEP = new(9090);
+
+service /testService on echoEP {
+
+    resource function get closureTest1(string status) returns string {
+        var addFunc = function (int funcInt) returns (string) {
+            return status;
+        };
+        return addFunc(1);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal:47]: We should probably try and add a test where this resource is invoked, right? @rdhananjaya, do we have such tests?

File: tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    boolean trees;
+    boolean night;
+|};
+
+listener test:MockListener echoEP = new(9090);
+
+service /testService on echoEP {
+
+    resource function get closureTest1(string status) returns string {
+        var addFunc = function (int funcInt) returns (string) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal:44]:
```suggestion
        var addFunc = function (int funcInt) returns string {
```

Can fix separately. We also usually use underscores in bal file names.

**Review by MaryamZi — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal (hunk 1 above)
Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal:47]: https://github.com/ballerina-platform/ballerina-lang/blob/fe42ba2cf2ed315f6aca8c8f1431276282384bba/tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/object/test_service_objects.bal#L88

Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal:47]: Also this
https://github.com/ballerina-platform/ballerina-lang/blob/fe42ba2cf2ed315f6aca8c8f1431276282384bba/tests/jballerina-unit-test/src/test/resources/test-src/services/service_decl.bal#L208

Comment (grainier) [reply tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal]: @MaryamZi, @rdhananjaya changed the test to have resource invocation (42c1683).

File: tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal (hunk 2 above)
Comment (grainier) [reply tests/jballerina-unit-test/src/test/resources/test-src/closures/closures-in-resource.bal]: Fixed (42c1683)

**Review by MaryamZi — APPROVED** [review]

**Review by rdhananjaya — APPROVED** [review]

**Pull request merged**